File: projects/gnrcore/packages/test/resources/grouplets/transforms/dati_pagamento.py
from gnr.web.gnrbaseclasses import BaseComponent


class Grouplet(BaseComponent):
    js_requires = 'grouplets/transforms/dati_pagamento'

    def __info__(self):
        return dict(
            caption='Dati Pagamento',
            priority=5,
            onLoading="function(data){dati_pagamento.onLoading(data);}",
            onSaving="function(data, sourceBag){dati_pagamento.onSaving(data, sourceBag);}"
        )

    def grouplet_main(self, pane, **kwargs):
        fb = pane.formlet(cols=2)
        fb.filteringSelect(
            value='^.CondizioniPagamento',
            values='TP01:Pagamento a rate,TP02:Pagamento completo,TP03:Anticipo',
            lbl='Condizioni Pagamento',
            colspan=2,
            width='100%'
        )
        pane.bagGrid(
            frameCode='dettaglio_pagamento',
            storepath='#FORM.record.DettaglioPagamento',
            struct=self._dati_pagamento_struct,
            datapath='#FORM.grid_dettaglio_pagamento',
            title='Dettaglio Pagamento',
            height='200px',
            addrow=True,
            delrow=True
        )

    def _dati_pagamento_struct(self, struct):
        r = struct.view().rows()
        r.cell('ModalitaPagamento', name='Modalita',
               width='15em',
               edit=dict(tag='filteringSelect',
                         values=','.join([
                             'MP01:Contanti',
                             'MP02:Assegno',
                             'MP05:Bonifico',
                             'MP08:Carta di pagamento',
                             'MP12:RIBA'
                         ])))
        r.cell('DataScadenzaPagamento', name='Data Scadenza',
               width='10em', dtype='D',
               edit=dict(tag='dateTextBox'))
        r.cell('ImportoPagamento', name='Importo',
               width='10em', dtype='N',
               edit=True,
               format='###,###,##0.00')

File: projects/gnrcore/packages/test/resources/grouplets/transforms/dati_pagamento.js
const dati_pagamento = {
    onLoading: function(data) {
        const nodes = data.getNodes();
        const gridBag = new gnr.GnrBag();
        let i = 0;
        const toRemove = [];
        for (const [k, node] of nodes.entries()) {
            if (node.label === 'DettaglioPagamento') {
                const val = node.getValue();
                if (val instanceof gnr.GnrBag) {
                    gridBag.setItem('r_' + i, val.deepCopy());
                } else {
                    gridBag.setItem('r_' + i, new gnr.GnrBag(val));
                }
                toRemove.push(k);
                i++;
            }
        }
        for (const idx of toRemove.reverse()) {
            data.delItem('#' + idx);
        }
        if (i > 0) {
            data.setItem('DettaglioPagamento', gridBag);
        }
    },

    onSaving: function(data, sourceBag) {
        const dettaglio = data.getItem('DettaglioPagamento');
        if (!dettaglio) {
            return;
        }
        data.delItem('DettaglioPagamento');
        const nodes = sourceBag.getNodes();
        for (let k = nodes.length - 1; k >= 0; k--) {
            if (nodes[k].label === 'DettaglioPagamento') {
                sourceBag.delItem('#' + k);
            }
        }
        for (const n of dettaglio.getNodes()) {
            const val = n.getValue();
            if (val instanceof gnr.GnrBag) {
                sourceBag.addItem('DettaglioPagamento', val.deepCopy());
            } else {
                sourceBag.addItem('DettaglioPagamento', new gnr.GnrBag(val));
            }
        }
    }
};
